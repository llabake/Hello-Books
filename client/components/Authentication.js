import React, { Component } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import { Redirect } from 'react-router-dom';
import toastMessage from '../helpers/toastMessage';

/**
 * 
 * 
 * @export
 * @param {any} Component to be rendere
 * @returns 
 */
export default function requireAuthentication(Component) {
  /**
   * 
   * 
   * @class Authenticate
   * @extends {Component}
   */
  class Authenticate extends Component {
    static propTypes = {
      authenticated: PropTypes.bool.isRequired,
      dispatch: PropTypes.func.isRequired
    };
  /**
   * 
   * 
   * @returns { js } protected component 
   * @memberof Authenticate
   */
  render() {
      return (
        <div>
          {this.props.authenticated
            ? <Component {...this.props} />
            :
            <Redirect to={{
              pathname: '/signin',
              state: { from: this.props.location }
            }}
            />
          }
          {/* toastMessage(' Please login to access this route', 'failure') */}
          
        </div>
      );
    }
  }

  const mapStateToProps = ({ userReducer }) => ({
    authenticated: userReducer.authenticated,
  });

  return connect(mapStateToProps)(Authenticate);
}
