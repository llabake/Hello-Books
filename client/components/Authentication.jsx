import React, { Component } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import { Redirect } from 'react-router-dom';
import toastMessage from '../helpers/toastMessage';
import { authenticateUser } from '../helpers/utils';

/**
 * 
 * 
 * @export
 * 
 * @param {any} ComposedComponent the child component to be rendered
 * 
 * @returns {void}
 */
export default function requireAuthentication(ComposedComponent) {
  /**
   * 
   * 
   * @class Authenticate
   * 
   * @extends {Component}
   */
  class Authenticate extends Component {
    static propTypes = {
      // authenticated: PropTypes.bool.isRequired,
      dispatch: PropTypes.func.isRequired
    };
    /**
     * 
     * 
     * @param {any} nextProps
     * 
     * @memberof Authenticate
     * 
     * @returns {void}
     */
    componentWillReceiveProps(nextProps) {
      if (!nextProps.isAuthenticated) {
        <Redirect to={{
          pathname: '/signin',
          state: { from: this.props.location }
        }}
        />
      }
    }
    /**
     * 
     * 
     * @returns { object } returns the protected object
     * 
     * @memberof Authenticate
     */
    render() {
      const { isAuthenticated } = authenticateUser()
      return (
        <div>
          {isAuthenticated
            ? <ComposedComponent {...this.props} />
            :
            <Redirect to={{
              pathname: '/signin',
              state: { from: this.props.location }
            }}
            />
          }
          {/* toastMessage(' Please login to access this route', 'failure') */}

        </div>
      );
    }
  }


  return connect(mapStateToProps, null)(Authenticate);
}

export const mapStateToProps = (state) => {
  return {
    error: state.userReducer.error,
    user: state.userReducer.authUser,
    authenticated: state.userReducer.authenticated
  };
};