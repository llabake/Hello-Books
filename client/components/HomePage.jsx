import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { Link } from 'react-router-dom';
import { connect } from 'react-redux'

import Slider from '../components/common/Slider';
import Carousel from '../components/common/Carousel';
import BookCard from '../components/common/BookCard';
import { fetchPopularBooks } from '../actions/bookAction';

import leanstart from '../media/lean start.jpg';
import ajaxLoader from '../media/ajax-loader.gif';


/**
 * 
 * 
 * @export
 * @class HomePage
 * @extends {Component}
 */
class HomePage extends Component {
  // static propTypes = {
  //   ,
  // }
  /**
   * Creates an instance of HomePage.
   * @param {any} props
   * 
   * @memberof HomePage
   */
  constructor(props) {
    super(props);
    this.state = {};
  }


  /**
   * @returns {Array} most popular books array
   * 
   * @memberof HomePage
   */
  componentWillMount() {
    this.props.fetchPopularBooks();
  }

  /**
   * 
   * @returns {function} initialises the side bav
   * 
   * @memberof HomePage
   */
  // componentDidMount() {
  //   $(".button-collapse").sideNav();
  // }




  /**
  * 
  * 
  * @returns {Object} html
  * 
  * @memberof HomePage
  */
  render() {
    const { popularBooks, loadingPopularBooks } = this.props;
    return (
      <div>
        <Slider />
        <div >
          <h4 className="center-align">
            <span>Most Popular BOOKS</span>
          </h4>
        </div>
        <div className="divider" />
        {loadingPopularBooks ? 
          <div className="center-align" style={{ marginTop: '1em', marginBottom: '1em' }}>
            <img src={ajaxLoader} alt="Loading..." />
          </div> : ''}
        <div className="section">
          <div className="container">
            <div className="row">
              {!loadingPopularBooks ? popularBooks.length ?
                popularBooks.map((book) => {
                  return <div key={book.id}><BookCard book={book} /></div>
                })
                :
                <h5 className="center-align"> No books found </h5>
                : ''}
            </div>
          </div>
        </div>
        <div className="divider" />
        <div className="section similar">
          <h3 className="center-align">
            Based on readers favorites
          </h3>
          <div><Carousel /></div>
        </div>
      </div>
    );
  }
}

const mapStateToProps = (state) => {
  return {
    popularBooks: state.bookReducer.popularBooks,
    loadingPopularBooks: state.bookReducer.loadingPopularBooks
  };
};


const mapDispatchToProps = (dispatch) => {
  return {
    fetchPopularBooks: () => dispatch(fetchPopularBooks()),
  };
};


HomePage.propTpyes = {
  fetchPopularBooks: PropTypes.func,
  popularBooks: PropTypes.arrayOf(PropTypes.object)
}
export default connect(mapStateToProps, mapDispatchToProps)(HomePage);
