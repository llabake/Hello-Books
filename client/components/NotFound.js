import React from 'react';
import { Link } from 'react-router-dom';
import SearchBar from './common/SearchBar';

const NotFound = (props) =>
  <div className="container" style= {{ height: "95vh"}}>
  <div style={{marginLeft: "0px"}}>
    <h4>This is embarassing, something's missing.</h4>
    <b>404: page not found</b>
  </div>
    <div style={{margin: "10 55px 20 35px",}}>
      <SearchBar />
    </div>
    <div >
      <ul style={{"textAlign": "center"}}>
      <li style={{"display": "block"}}><Link to="/">Home</Link>
      </li>
      <li style={{"display": "block"}}><Link to="/allbooks">Books</Link>
      </li>
      <li style={{"display": "block"}}><Link to="/profile">Profile</Link>
      </li>
      </ul>
    </div>
  </div>

export default NotFound;

