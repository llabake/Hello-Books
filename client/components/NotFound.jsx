import React from 'react';
import { Link } from 'react-router-dom';
import SearchBar from './common/SearchBar';

const NotFound = (props) =>
  <div className="container">
    <div style={{ marginLeft: "0px" }} id="not-found">
      <h4>This is embarassing, something's missing.</h4>
      <b>404: page not found</b>
    </div>
    <div>
      <SearchBar />
    </div>
    <div >
      <ul style={{ "textAlign": "center" }}>
        <li style={{ "display": "block" }}><Link to="/">Home</Link>
        </li>
        <li style={{ "display": "block" }}><Link to="/allbooks">Books</Link>
        </li>
        <li style={{ "display": "block" }}><Link to="/profile">Profile</Link>
        </li>
      </ul>
    </div>
  </div>

export default NotFound;

