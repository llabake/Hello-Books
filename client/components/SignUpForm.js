import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { Redirect, Link } from 'react-router-dom';
import { connect } from 'react-redux'

import { checkUserExist, signUpUser } from '../actions/userAction';
import TextInput from '../components/common/TextInput';
import inputValidator from '../helpers/inputValidator';

/**
 * 
 * 
 * @export
 * @class SignUpForm
 * @extends {Component}
 */
class SignUpForm extends Component {
  /**
   * Creates an instance of SignUpForm.
   * @param {any} props 
   * @memberof SignUpForm
   */
  constructor(props) {
    super(props);
    this.state = {
      firstName: '',
      lastName: '',
      username: '',
      email: '',
      password: '',
      confirmPassword: '',
      errors: {}, 
      saving: false,
      isValid: false,
      userExist: {},

    };
    this.handleChange = this.handleChange.bind(this);
    this.handleSubmit = this.handleSubmit.bind(this);
    this.handleBlur =this.handleBlur.bind(this);
  }

  /**
   * 
   * 
   * @param {any} event 
   * @memberof SignUpForm
   * @returns {Object} object containing state
   */
  handleChange (event) {
    this.setState({
      [event.target.name]: event.target.value
    }, () => this.validate());
  }

  /**
   * 
   * 
   * @param {any} event 
   * @memberof SignUpForm
   * @returns {Object} object containing user detail
   */
  handleSubmit(event) {
    event.preventDefault();
    const userData = this.state;
    if(this.validate()) {
      this.setState({
        error: {},
        saving: true
      });
      this.props.signUpUser(userData)
    }
  }

   /**
   * 
   * 
   * @memberof SignUpForm
   * 
   * @returns {void}
   */
  componentWillMount() {
    if (this.props.authenticated) {
      this.props.history.push('/profile');
    }
  }
  /**
   * 
   * 
   * @param {any} nextProps 
   * @memberof SignUpForm
   * 
   * @returns {Object} User onbject
   */
  componentWillReceiveProps(nextProps) {
    if(nextProps.authenticated) {
      setTimeout(() => {
        this.props.history.push('/allbooks')
      }, 2000)
    } else {
      this.setState({ saving: false })
    }
  }




  /**
   * @returns {obejct} user validation status
   * 
   * @param {any} event 
   * @memberof SignUpForm
   */
  handleBlur(event) {
    const field = event.target.name;
    const userInput = event.target.value;
    if(userInput !== '') {
      this.props.checkUserExist(field, userInput)
      .then(() => {
        this.setState({ 
          userExist: {}
        }, () => {
          this.validate()
        })
      })
      .catch((error) => {
        const errors = this.state.errors;
        const userExist = this.state.userExist;
        errors[field].push(error.response.data.message)
        userExist[field] = error.response.data.message
        this.setState({ errors, userExist })
      })
    }
  }

  /**
   * @returns {Object} object containing validation status and corresponding errors
   * 
   * @memberof SignUpForm
   */
  validate() {
    const { errors, isValid } = inputValidator.signUp(this.state);
    this.setState({ isValid, errors });
    return isValid;
  }

  /**
   * 
   * 
   * @returns {Object} object containing user detail
   * @memberof SignUpForm
   */
  render () {  
    const { errors, isValid, saving, } = this.state;
    return (
      <div>
        <div id="banner">
          <div className="container form-style">
            <div className="row">
              <div className="row signup-head">  
                <div className="col s12"> 
                  <h5 className="center-align"> Create an HelloBooks Account </h5> 
                </div>
              </div> 
              <div className="col s12 ">
                <div className="row ">
                  <form className="col s12 signup" onSubmit={this.handleSubmit}>      
                    <div className="s12">
                      <TextInput
                      id = 'first_name'
                      type = 'text'
                      icon = 'account_circle'
                      name = 'firstName'
                      placeholder = 'First Name'
                      onChange = {this.handleChange}
                      value = {this.state.firstName}
                      errors = {errors.firstName}
                      />
                      <TextInput
                      id = 'last_name'
                      type = 'text'
                      icon = 'account_circle'
                      name = 'lastName'
                      placeholder = 'Last Name'
                      onChange = {this.handleChange}
                      clearError={this.clearError}
                      value = {this.state.lastName}
                      errors = {errors.lastName}
                      />
                    </div>
                    <TextInput
                    id = 'username'
                    type = 'text'
                    icon = 'account_circle'
                    name = 'username'
                    placeholder = 'Username'
                    onChange = {this.handleChange}
                    onBlur = {this.handleBlur}
                    value = {this.state.username}
                    errors = {errors.username}
                    />
                    <TextInput
                    id = 'email'
                    type = 'email'
                    icon = 'email'
                    name = 'email'
                    placeholder = 'Email'
                    onChange = {this.handleChange}
                    onBlur = {this.handleBlur}
                    value = {this.state.email}
                    errors = {errors.email}
                    />
                    <TextInput
                    id = 'password'
                    type = 'password'
                    icon = 'lock'
                    name = 'password'
                    placeholder = 'Password'
                    onChange = {this.handleChange}
                    value = {this.state.password}
                    errors = {errors.password}
                    />
                    <TextInput
                    id = 'confirm-password'
                    type = 'password'
                    icon = 'lock'
                    name = 'confirmPassword'
                    placeholder = 'Re-enter Password'
                    onChange = {this.handleChange}
                    value = {this.state.confirmPassword}
                    errors = {errors.confirmPassword}
                    />
                    <button type="submit" className="waves-effect waves-light btn" disabled= {!isValid || saving}>Register</button>
                  </form>
                </div>
              </div>
            </div>
            <div className="terms">
              {/* <p>By creating an account you agree to our <Link to="#">Terms & Privacy</Link>.</p> */}
              <p> Already have an account? <Link to="/signin">Login Now</Link></p> 
            </div>
          </div> 
        </div>
      </div>


    );
  }
}

const mapStateToProps = (state) => {
  return {
    errors: state.errors,
    user: state.userReducer.authUser,
    authenticated: state.userReducer.authenticated    
  };
};

const mapDispatchToProps = (dispatch) => {
  return {
    checkUserExist: (field, userInput) => dispatch(checkUserExist(field, userInput)),
    signUpUser: (userData) => dispatch(signUpUser(userData))
  };
};

export default connect(mapStateToProps, mapDispatchToProps)(SignUpForm);

