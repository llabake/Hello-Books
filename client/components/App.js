import React, { Component } from 'react';
import { render } from 'react-dom';
import { BrowserRouter as Router, Route, Switch } from 'react-router-dom';
import requireAuthentication from './Authentication';
import HomePage from './HomePage';
import SignUpForm from './SignUpForm';
import SignInForm from './SignInForm';
import AllBooks from './main/AllBooks'
import AddBook from './adminDashBoard/AddBook';
import SingleBook from './main/SingleBook';
import AdminDashBoard from './adminDashBoard/AdminDashBoard';
import UserProfilePage from './userDashBoard/UserProfilePage';
import Favorite from './userFavorite/Favorite';
import SearchResult from './common/SearchResult';
import EditProfile from './userDashBoard/EditProfile';
import NotFound from './NotFound';

import '../styles/style.scss';
import Header from './common/Header';
import Footer from './common/Footer';
/**
 * 
 * 
 * @export
 * @class App
 * @extends {Component}
 */
export default class App extends Component {

  componentDidMount() {
    $('.button-collapse').sideNav({
      menuWidth: 250, // Default is 300
      edge: 'left', // Choose the horizontal origin
      draggable: true // Choose whether you can drag to open on touch screens
    }
  );
  // $('.carousel').carousel({
  //   dist: 0, // Perspective zoom. 
  //   shift: 0, // Set the spacing of the center item.
  //   padding: 50, // Set the padding between non center items.
  //   indicators: true, // Show indicators. Default is false
  // });
  // autoplay()
  // function autoplay() {
    // $('.carousel').carousel('next');
    // setTimeout(autoplay, 4500)}

    $('.slider').slider();
    $('.dropdown-button').dropdown({
      belowOrigin: true,
    });
    $('.tooltipped').tooltip({ delay: 50 });
    $('.tabs').tabs();
    $('.tooltipped').tooltip();
  

    // $('.modal').modal();
    // $('select').material_select();
    
  }
  /**
   * 
   * 
   * @returns {object} main App
   * @memberof App
   */
  render () {
    return (
      <Router>
        <div className="body">
        <header>
        <Header />
        </header>
        <main>
        <Switch>
          <Route exact path='/' component={HomePage} />
          <Route exact path='/signup' component={SignUpForm} />
          <Route exact path='/signin' component={SignInForm} />
          <Route exact path='/allbooks' component={requireAuthentication(AllBooks)} />
          <Route exact path='/addbook' component={requireAuthentication(AddBook)} />
          <Route exact path='/admindashboard' component={requireAuthentication(AdminDashBoard)} />
          <Route exact path='/book/:bookId' component={requireAuthentication(SingleBook)} />
          <Route exact path='/profile' component={requireAuthentication(UserProfilePage)} />
          <Route exact path='/favorite' component={requireAuthentication(Favorite)}/>
          <Route exact path='/search/:searchTerm' component={requireAuthentication(SearchResult)}/>
          <Route exact path='/editprofile' component={requireAuthentication(EditProfile)} />
          <Route component={NotFound} />
        </Switch>
        </main>
        <footer>
        <Footer />
        </footer>
        </div>
      </Router>
    );
  }
}





