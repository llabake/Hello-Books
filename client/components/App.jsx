import React, { Component } from 'react';
import { render } from 'react-dom';
import { BrowserRouter as Router, Route, Switch } from 'react-router-dom';
import requireAuthentication from './Authentication';
import HomePage from './HomePage';
import SignUpForm from './SignUpForm';
import SignInForm from './SignInForm';
import AllBooks from './main/AllBooks'
import AddBook from './adminDashBoard/AddBook';
import SingleBook from './main/SingleBook';
import AdminDashBoard from './adminDashBoard/AdminDashBoard';
import UserProfilePage from './userDashBoard/UserProfilePage';
import Favorite from './userFavorite/Favorite';
import SearchResult from './common/SearchResult';
import EditProfile from './userDashBoard/EditProfile';
import NotFound from './NotFound';

import '../styles/style.scss';
import Header from './common/Header';
import Footer from './common/Footer';
/**
 * 
 * 
 * @export
 * @class App
 * @extends {Component}
 */
export default class App extends Component {
  /**
   *@returns {void}
   *
   * @memberof App
   */
  componentDidMount() {
    $('.button-collapse').sideNav({
      menuWidth: 250, // Default is 300
      edge: 'left', // Choose the horizontal origin
      draggable: true // Choose whether you can drag to open on touch screens
    }
    );
    $('.slider').slider({
      full_width: true,
    }
    );
    $('.dropdown-button').dropdown({
      belowOrigin: true,
    });
    $('.tooltipped').tooltip({ delay: 50 });
    $('ul.tabs').tabs();
    $('.tooltipped').tooltip({ delay: 50 });
  }

  /**
   * 
   * 
   * @returns {object} main App
   * @memberof App
   */
  render() {
    return (
      <Router>
        <div className="body">
          <header>
            <Header />
          </header>
          <main>
            <Switch>
              <Route exact path='/' component={HomePage} />
              <Route exact path='/signup' component={SignUpForm} />
              <Route exact path='/signin' component={SignInForm} />
              <Route exact path='/allbooks' component={requireAuthentication(AllBooks)} />
              <Route exact path='/addbook' component={requireAuthentication(AddBook)} />
              <Route exact path='/admindashboard' component={requireAuthentication(AdminDashBoard)} />
              <Route exact path='/book/:bookId' component={requireAuthentication(SingleBook)} />
              <Route exact path='/profile' component={requireAuthentication(UserProfilePage)} />
              <Route exact path='/favorite' component={requireAuthentication(Favorite)} />
              <Route exact path='/search/:searchTerm' component={requireAuthentication(SearchResult)} />
              <Route exact path='/editprofile' component={requireAuthentication(EditProfile)} />
              <Route component={NotFound} />
            </Switch>
          </main>
          <footer>
            <Footer />
          </footer>
        </div>
      </Router>
    );
  }
}





