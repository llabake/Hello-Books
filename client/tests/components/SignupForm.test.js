import React from 'react';
import { SignUpForm, mapDispatchToProps, mapStateToProps } from '../../components/SignUpForm';

const setup = () => {
  const props = {
    authenticated: false,
    dispatch: jest.fn(),
    handleSubmit: jest.fn(),
    handleChange: jest.fn(),
    handleBlur: jest.fn(),
    signUpUser: jest.fn(),
    history: {
      push: jest.fn()
    },
    checkUserExist: jest.fn().mockImplementation(() => Promise.resolve(1))
  };
  const wrapper = shallow(<SignUpForm {...props} />);

  return {
    wrapper,
    props
  };
}

describe('Signup Page component', () => {
  test('should render correctly', () => {
    const { wrapper } = setup();
    expect(wrapper).toMatchSnapshot();
  });
  test('should render form for user input sction', () => {
    const { wrapper } = setup();
    expect(wrapper.find('form').exists()).toBe(true);
  });
  test('should have six input fields', () => {
    const { wrapper } = setup();
    expect(wrapper.find('TextInput').length).toEqual(6);
  });
  test('should have state of all sign up form fields to be an empty string', () => {
    const { wrapper } = setup();
    expect(wrapper.state().firstName).toBe('');
    expect(wrapper.state().lastName).toBe('');
    expect(wrapper.state().username).toBe('');
    expect(wrapper.state().email).toBe('');
    expect(wrapper.state().password).toBe('');
    expect(wrapper.state().confirmPassword).toBe('');
  });

  test('should return errors when trying to submit an empty form', () => {
    const { wrapper } = setup();

    wrapper.find('form').simulate('submit', { preventDefault: jest.fn() });
    expect(wrapper.state().isValid).toBe(false);
    expect(wrapper.state().saving).toBe(false)
  });

  test('should redirect to previous page or home if user is already authenticated', () => {
    const { wrapper } = setup();
    wrapper.setProps({ authenticated: true });
    expect(wrapper).toMatchSnapshot();
  });


  test('calls handle change when first name changes', () => {
    const { wrapper, props } = setup();
    const event = { target: { value: 'first name', name: 'firstName' } };

    wrapper.find('#first_name').simulate('change', event);
    expect(wrapper.instance().state.firstName).toEqual('first name')
  });

  test('calls onSubmit', () => {
    const signUpUser = jest.fn();
    const handleSubmit = jest.fn();
    const { wrapper, props } = setup();
    const event = {
      preventDefault: jest.fn(),
      target: {}
    };
    const fields = {
      usernameField: wrapper.find("#username"),
      passwordField: wrapper.find("#password"),
      firstnameField: wrapper.find("#firstName"),
      lastnameField: wrapper.find('#lastName'),
      emailField: wrapper.find("#email"),
      signUpForm: wrapper.find('.signup'),
    }

    const { usernameField, passwordField, firstnameField, lastnameField, emailField, confirmPasswordField, signUpForm } = fields
    usernameField.simulate('change', { target: { name: 'username', value: 'username' } });
    passwordField.simulate('change', { target: { name: 'password', value: 'username' } });
    emailField.simulate('change', { target: { name: 'email', value: 'xyz@mail.com' } });

    expect(wrapper.instance().state.firstName).toBe('');
    expect(wrapper.instance().state.lastName).toBe('');
    expect(wrapper.instance().state.saving).toBe(false);
    expect(wrapper.instance().state.isValid).toBe(false);
    expect(wrapper.instance().state.confirmPassword).toBe('');
    expect(wrapper.instance().state.errors.firstName)
      .toEqual(['firstName is required']);
    expect(wrapper.instance().state.errors.lastName)
      .toEqual(['lastName is required']);
    expect(wrapper.instance().state.errors.confirmPassword)
      .toEqual(['Please confirm your password']);
  });

  test('should handle form submission when the form inputs are valid', () => {
    const { wrapper, props, state } = setup();
    const event = {
      preventDefault: jest.fn(),
      target: {}
    }
    wrapper.setState({
      username: 'username',
      password: 'P@ssword1',
      firstName: 'firstName',
      confirmPassword: 'P@ssword1',
      email: 'email@email.com',
      lastName: "lastName"
    });
    props.signUpUser.mockReset();
    wrapper.instance().handleSubmit(event)
    expect(wrapper.instance().state.username).toBe('username');
    expect(props.signUpUser).toHaveBeenCalled()
  });


  test('should call user exist to check if user exists before sign up', () => {
    const { wrapper, props, state } = setup();
    const handleblurSpy = jest.spyOn(SignUpForm.prototype, 'handleBlur')
    const event = {
      preventDefault: jest.fn(),
      target: {
        name: 'username',
        value: 'username '
      }
    }
    wrapper.setState({
      username: 'username',
      // userExist: {
      //   username: 'username error something'
      // }
    });
    wrapper.instance().handleBlur(event)
    expect(wrapper.instance().state.username).toBe('username');
    expect(props.checkUserExist).toHaveBeenCalled()
  });

  test('should return an object with props required by component from store', () => {
    const store = {
      authenticated: false,
      user: {}
    };
    expect(mapStateToProps({ userReducer: {} })).toHaveProperty('authenticated');
    expect(mapStateToProps({ userReducer: {} })).toHaveProperty('user');
  });

  test('should dispatch needed actions', () => {
    const dispatch = jest.fn();
    expect(mapDispatchToProps(dispatch)).toHaveProperty('signUpUser')
    expect(mapDispatchToProps(dispatch)).toHaveProperty('checkUserExist')
    const { checkUserExist, signUpUser } = mapDispatchToProps(dispatch);
    signUpUser();
    expect(dispatch).toHaveBeenCalled();
    checkUserExist();
    expect(dispatch).toHaveBeenCalled();
  })

  test('should call componentWillMount', () => {
    let spy;
    afterEach(() => {
      spy.mockClear();
    });

    const props = {
      authenticated: true,
      dispatch: jest.fn(),
      handleSubmit: jest.fn(),
      handleChange: jest.fn(),
      handleBlur: jest.fn(),
      history: {
        push: jest.fn()
      },
    };
    spy = jest.spyOn(SignUpForm.prototype, 'componentWillMount');
    const wrapper = shallow(<SignUpForm {...props} />)
    // const route = wrapper.find({ path: '/profile' });

    expect(SignUpForm.prototype.componentWillMount).toHaveBeenCalledTimes(1)
    expect(spy).toHaveBeenCalled();
    // route.props().render();
    // expect(route).toBeTruthy();
  })

  test('should call componentWillReceiveProps', () => {
    let spy;
    afterEach(() => {
      spy.mockClear();
    });

    const props = {
      authenticated: false,
      dispatch: jest.fn(),
      handleSubmit: jest.fn(),
      handleChange: jest.fn(),
      handleBlur: jest.fn(),
      history: {
        push: jest.fn()
      },
    };

    const nextProps = {
      authenticated: true,
      dispatch: jest.fn(),
      handleSubmit: jest.fn(),
      handleChange: jest.fn(),
      handleBlur: jest.fn(),
      history: {
        push: jest.fn()
      },
    }
    spy = jest.spyOn(SignUpForm.prototype, 'componentWillReceiveProps');
    const wrapper = shallow(<SignUpForm {...props} />)
    wrapper.setProps(nextProps);
    expect(wrapper.instance().componentWillReceiveProps).toHaveBeenCalledTimes(1)
  })

  test('should call componentWillReceiveProps and set saving to be false', () => {
    let spy;
    afterEach(() => {
      spy.mockClear();
    });

    const props = {
      authenticated: false,
      dispatch: jest.fn(),
      handleSubmit: jest.fn(),
      handleChange: jest.fn(),
      handleBlur: jest.fn(),
      history: { pathname: '' }
    };

    const nextProps = {
      authenticated: false,
      dispatch: jest.fn(),
      handleSubmit: jest.fn(),
      handleChange: jest.fn(),
      handleBlur: jest.fn(),
      history: {
        push: jest.fn()
      },
    }
    spy = jest.spyOn(SignUpForm.prototype, 'componentWillReceiveProps');
    const wrapper = shallow(<SignUpForm {...props} />)
    wrapper.setProps(nextProps);
    expect(SignUpForm.prototype.componentWillReceiveProps).toHaveBeenCalled()
    expect(spy).toHaveBeenCalled();

  })
});
