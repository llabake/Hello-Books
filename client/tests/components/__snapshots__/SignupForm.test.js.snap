// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Signup Page component should redirect to previous page or home if user is already authenticated 1`] = `
<div>
  <div
    id="banner"
  >
    <div
      className="container form-style"
    >
      <div
        className="row"
      >
        <div
          className="row signup-head"
        >
          <div
            className="col s12"
          >
            <h5
              className="center-align"
            >
               Create an HelloBooks Account 
            </h5>
          </div>
        </div>
        <div
          className="col s12 "
        >
          <div
            className="row "
          >
            <form
              className="col s12 signup"
              onSubmit={[Function]}
            >
              <div
                className="s12"
              >
                <TextInput
                  icon="account_circle"
                  id="first_name"
                  name="firstName"
                  onChange={[Function]}
                  placeholder="First Name"
                  type="text"
                  value=""
                />
                <TextInput
                  icon="account_circle"
                  id="last_name"
                  name="lastName"
                  onChange={[Function]}
                  placeholder="Last Name"
                  type="text"
                  value=""
                />
              </div>
              <TextInput
                icon="account_circle"
                id="username"
                name="username"
                onBlur={[Function]}
                onChange={[Function]}
                placeholder="Username"
                type="text"
                value=""
              />
              <TextInput
                icon="email"
                id="email"
                name="email"
                onBlur={[Function]}
                onChange={[Function]}
                placeholder="Email"
                type="email"
                value=""
              />
              <TextInput
                icon="lock"
                id="password"
                name="password"
                onChange={[Function]}
                placeholder="Password"
                type="password"
                value=""
              />
              <TextInput
                icon="lock"
                id="confirm-password"
                name="confirmPassword"
                onChange={[Function]}
                placeholder="Re-enter Password"
                type="password"
                value=""
              />
              <button
                className="waves-effect waves-light btn"
                disabled={true}
                type="submit"
              >
                Register
              </button>
            </form>
          </div>
        </div>
      </div>
      <div
        className="terms"
      >
        <p>
           Already have an account? 
          <Link
            replace={false}
            to="/signin"
          >
            Login Now
          </Link>
        </p>
      </div>
    </div>
  </div>
</div>
`;

exports[`Signup Page component should render correctly 1`] = `
<div>
  <div
    id="banner"
  >
    <div
      className="container form-style"
    >
      <div
        className="row"
      >
        <div
          className="row signup-head"
        >
          <div
            className="col s12"
          >
            <h5
              className="center-align"
            >
               Create an HelloBooks Account 
            </h5>
          </div>
        </div>
        <div
          className="col s12 "
        >
          <div
            className="row "
          >
            <form
              className="col s12 signup"
              onSubmit={[Function]}
            >
              <div
                className="s12"
              >
                <TextInput
                  icon="account_circle"
                  id="first_name"
                  name="firstName"
                  onChange={[Function]}
                  placeholder="First Name"
                  type="text"
                  value=""
                />
                <TextInput
                  icon="account_circle"
                  id="last_name"
                  name="lastName"
                  onChange={[Function]}
                  placeholder="Last Name"
                  type="text"
                  value=""
                />
              </div>
              <TextInput
                icon="account_circle"
                id="username"
                name="username"
                onBlur={[Function]}
                onChange={[Function]}
                placeholder="Username"
                type="text"
                value=""
              />
              <TextInput
                icon="email"
                id="email"
                name="email"
                onBlur={[Function]}
                onChange={[Function]}
                placeholder="Email"
                type="email"
                value=""
              />
              <TextInput
                icon="lock"
                id="password"
                name="password"
                onChange={[Function]}
                placeholder="Password"
                type="password"
                value=""
              />
              <TextInput
                icon="lock"
                id="confirm-password"
                name="confirmPassword"
                onChange={[Function]}
                placeholder="Re-enter Password"
                type="password"
                value=""
              />
              <button
                className="waves-effect waves-light btn"
                disabled={true}
                type="submit"
              >
                Register
              </button>
            </form>
          </div>
        </div>
      </div>
      <div
        className="terms"
      >
        <p>
           Already have an account? 
          <Link
            replace={false}
            to="/signin"
          >
            Login Now
          </Link>
        </p>
      </div>
    </div>
  </div>
</div>
`;
